### PR TITLE
WEEK 7: Implement Policy Interceptor and Agent Profiles for FinBot Security Framework

### PR DESCRIPTION
**Pull request by Jean-Regis-M**
## Summary

This week's work focused on implementing the core security infrastructure for the FinBot project by developing the **Policy Interceptor** and predefined **Agent Profiles** as outlined in the GSoC proposal. The implementation addresses OWASP ASI02 (Tool Misuse), ASI03 (Excessive Agency), and ASI07 (Prompt Injection) through least-privilege access controls and intent-based security mechanisms.

### Core Implementation (`finbot/aegis/policy/`)

1. **`__init__.py`** - Package initialization exporting all policy components
2. **`agent_profiles.py`** - Defines six least-agency agent profiles:
   - Data Analyst: Read-only access to financial data tools
   - Report Generator: Limited write access for report generation
   - API Integrator: Controlled access to external API wrappers
   - Security Auditor: Specialized permissions for security scanning tools
   - Workflow Orchestrator: Privileges to manage and coordinate workflows
   - Research Assistant: Broad but monitored access for research tasks
3. **`interceptor.py`** - Implements `MCPPolicyInterceptor` with:
   - Real-time MCP tool call interception
   - Policy enforcement based on agent profiles
   - Resource limit enforcement (rate limiting)
   - Intent capsule creation (stubbed for Week 8)
   - Policy violation logging and metrics collection
   - Integration hooks for memory namespace provenance (stubbed for Week 8)

### Week 8 Stubs (Forward Compatibility)
- **`capsule.py`** - Stub for `IntentCapsule` with HMAC signing/validation interface
- **`memory.py`** - Stub for `MemoryNamespaceManager` with provenance tracking interface

### Comprehensive Test Suite
- **`tests/unit/aegis/test_policy_interceptor.py`** - Full test coverage including:
  - Tool authorization and blocking tests
  - Resource limit enforcement validation
  - Agent profile behavior verification
  - Metrics reporting accuracy
  - Policy violation logging checks

## Key Features

- **Least-Privilege Access Control**: Each agent type operates with precisely defined tool allowances
- **Dynamic Policy Enforcement**: Interceptor wraps tool execution with real-time policy checks
- **Rate Limiting**: Configurable resource limits prevent abuse and ensure fair usage
- **Audit Ready**: Comprehensive logging of all policy decisions for compliance
- **Extensible Design**: Clean separation of concerns with stubs prepared for Week 8 enhancements
- **Security Framework Alignment**: Directly addresses OWASP ASIS Top 10 for LLM Applications

## Testing Status

✅ All unit tests passing
✅ Edge cases covered (invalid profiles, exceeded limits, unauthorized tool access)
✅ Metrics collection verified
✅ Agent profile combinations tested

## Next Steps (Week 8)

In Week 8, the stub files will be replaced with full implementations featuring:
- Cryptographic HMAC signing and validation for intent capsules
- Secure memory namespace isolation with provenance tracking
- Tamper-evident audit log generation
- Integration with the policy interceptor for end-to-end security

## Acknowledgments

Special thanks to my mentors @mekaizen  and @steadhac  for their continuous guidance, insightful feedback, and unwavering support throughout this GSoC journey. Their expertise has been instrumental in shaping this security-focused implementation.

### CODE REVIEW

File: finbot/aegis/policy/agent_profiles.py [hunk 1]
```diff
@@ -0,0 +1,237 @@
...
+        "data:transform",
+        "data:summarize",
+        "filesystem:read",
+    ],
+    tool_limits={
+        "database:query": {"max_calls_per_minute": 30},
+        "data:visualize": {"max_calls_per_minute": 10},
+        "filesystem:read": {"max_calls_per_minute": 60},
+    },
+    secret_key="data-analyst-secret-key-change-in-production",
```
Comment (mekaizen) [inline finbot/aegis/policy/agent_profiles.py]: Multiple agent profiles define hardcoded default secrets such as:

Do not embed predictable default secrets in code. Load them from configuration or generate cryptographically secure secrets.

import secrets

secret_key = (
    settings.get("AEGIS_AGENT_SECRET")
    or secrets.token_hex(32)
)

or fail fast
secret = settings.get("AEGIS_AGENT_SECRET")

if not secret:
    raise RuntimeError(
        "AEGIS_AGENT_SECRET must be configured."
    )

secret_key = secret

Comment (Jean-Regis-M) [reply finbot/aegis/policy/agent_profiles.py]: Thank you for your review and suggestions, @mekaizen! I've updated the agent profiles to remove hardcoded secrets and instead load the secret key from configuration (with a fallback to a cryptographically generated secret for development). The changes are in line with the project's security best practices. Please let me know if you have any further feedback.
